### PR TITLE
Function space SBP operators

### PR DESCRIPTION
**Pull request by JoshuaLampert**
This is a very first draft for tackling #132. For the [construction of the FSBP](https://arxiv.org/abs/2405.08770) an optimization problem needs to be solved. I plan to use Optim.jl for this. Thus, I added a package extension to avoid adding Optim.jl as a direct dependency. In order to discuss some design choices, I already created this PR.

- In the current draft, I use the `MatrixDerivativeOperator` to represent the functional space operators, i.e. `FunctionSpaceOperator` just returns `MatrixDerivativeOperator`. This is easy, but I'm not sure if this is the best design choice on the long run, e.g., this way we "forget" about the `basis`. A more versatile alternative would be to define a new struct that holds a `MatrixDerivativeOperator` and potentially more such as the `basis` for potential later use. In this case, several methods (such as `mass_matrix`, `integrate`, `*`...) could just call the corresponding methods of the `MatrixDerivativeOperator`.

- Another question concerns the derivatives of the basis functions, which are needed for the construction process. From a user perspective it would be most convenient if they would not need to be specified explicitly. Do you think, we could use ForwardDiff.jl (i.e. the package extension consists of ForwardDiff.jl and Optim.jl) to compute the derivatives or would that mean too much overhead?

I would be interested in your opinion on these two points @ranocha and if the current approach looks reasonable to you.

### CODE REVIEW

**Comment by JoshuaLampert** [comment]
Thanks for the answers!

- I don't have any specific application in mind, but I thought I might be worth considering the alternatives before one realizes that a more flexible approach would be better. But your suggestion to document that the general interface will be implemented, but the concrete type may change in the future sounds like a good way.
- Ok, I will use ForwardDiff.jl for the derivatives.

File: src/function_space_operators.jl [hunk 1]
```diff
@@ -0,0 +1,49 @@
...
+          "  arXiv, arXiv:2405.08770v1.")
+  end
+end
+
+"""
+    FunctionSpaceOperator(basis, x_min, x_max, nodes, source)
+
+Construct a `MatrixDerivativeOperator` that represents a derivative operator in a function space.
+"""
+function FunctionSpaceOperator(basis::Vector{BasisFunctions},
```
[the remark this answers is not in the document]
Comment (JoshuaLampert) [reply src/function_space_operators.jl]: Yes, makes sense. I removed the type in [5f85d67](https://github.com/ranocha/SummationByPartsOperators.jl/pull/268/commits/5f85d67239eba859010d270eab29886c8062a380).

File: src/function_space_operators.jl [hunk 2]
```diff
@@ -0,0 +1,49 @@
...
+                               x_min::T, x_max::T, nodes::Vector{T},
+                               source::SourceOfCoefficients) where {BasisFunctions, T, SourceOfCoefficients}
+
+  weights, D = construct_function_space_operator(basis, x_min, x_max, nodes, source)
+  accuracy_order = length(basis) # TODO: this might needs to be adjusted
```
[the remark this answers is not in the document]
Comment (JoshuaLampert) [reply src/function_space_operators.jl]: I added `accuracy_order` as a kwarg with default 0. Since the value shouldn't affect the implementation, this should be fine, at least for now.

File: src/function_space_operators.jl [hunk 3]
```diff
@@ -0,0 +1,49 @@
...
+Construct a `MatrixDerivativeOperator` that represents a derivative operator in a function space.
+"""
+function FunctionSpaceOperator(basis_functions,
+                               x_min::T, x_max::T, nodes::Vector{T},
+                               source::SourceOfCoefficients;
+                               accuracy_order = 0) where {T, SourceOfCoefficients}
+
+  weights, D = construct_function_space_operator(basis_functions, x_min, x_max, nodes, source)
+  return MatrixDerivativeOperator(x_min, x_max, nodes, weights, D, accuracy_order, source)
+end
```
Comment (ranocha) [inline src/function_space_operators.jl:46]: We should document that users need to load `Optim` and `ForwardDiff` for this.

File: src/function_space_operators.jl [hunk 4]
```diff
@@ -0,0 +1,49 @@
...
+
+# This function is extended in the package extension SummationByPartsOperatorsOptimExt
+function construct_function_space_operator end
```
Comment (ranocha) [inline src/function_space_operators.jl:49]: Would it also work if we defined the actual constructor only in the extension?

File: src/function_space_operators.jl [hunk 5]
```diff
@@ -0,0 +1,49 @@
...
+          "  arXiv, arXiv:2405.08770v1.")
+  end
+end
+
+"""
+    FunctionSpaceOperator(basis_functions, x_min, x_max, nodes, source; accuracy_order = 0)
+
+Construct a `MatrixDerivativeOperator` that represents a derivative operator in a function space.
+"""
+function FunctionSpaceOperator(basis_functions,
```
Comment (ranocha) [inline src/function_space_operators.jl:39]: A capitalized name suggests that it's a constructor returning something of type `FunctionSpaceOperator`, which is not the case. Maybe switch to `function_space_operator` or so? And document that it's not necessarily returning a `MatrixDerivativeOperator` as discussed earlier?

**Review by ranocha — COMMENTED** [review]
Nice!

> Up to now, I have focused on the "make it work" phase by trying to be as close as possible to the mathematical formulation. The "make it fast"(er) phase is yet to come. The optimization process is still taking too long for higher number of nodes, but there is probably a lot of potential for performance improvements. Before I dig deeper into some performance improvement, I would like to see your feedback @ranocha. Does this look ok to you in general or do you think the implementation needs some major alterations?

I think the general approach is fine. I just have some high-level comments right now. The lower-level implementation details could be touched later - we can also discuss this together.

> Up to now I have only added a few tests to cover the most basic. I had a brief look in SBP_operators_test.jl, but was not quite sure, which of these tests also make sense in this case. (I also didn't find any tests like @test M * D.D + D.D' * M ≈ B, which I found a bit surprising as this would have been one of the first things I would like to see tested for the SBP operators 😅)

Fair enough - feel free to add tests like this to the other files 😅 

> I have also noticed that the optimization needs a rather small tolerance for the gradient to be able to produce accurate results. I was also wondering about whether we should provide the user some possibility to influence the optimization, e.g., by allowing an additional kwarg options that is passed as Optim.Options to the optimization (with some reasonable default value), or by allowing to use a custom sigmoid function.

That sounds reasonable.

File: src/function_space_operators.jl (hunk 4 above)
Comment (JoshuaLampert) [reply src/function_space_operators.jl]: Like in [d4266fe](https://github.com/ranocha/SummationByPartsOperators.jl/pull/268/commits/d4266fe62f3306c0912913be1259b2e1512f9a52)? That also works, yes.

File: src/function_space_operators.jl (hunk 5 above)
Comment (JoshuaLampert) [reply src/function_space_operators.jl]: Fixed in [a2ef507](https://github.com/ranocha/SummationByPartsOperators.jl/pull/268/commits/a2ef507933aa4964d4264d1a86427f0242d519b9).

File: src/function_space_operators.jl (hunk 3 above)
Comment (JoshuaLampert) [reply src/function_space_operators.jl]: I added a sentence to the docstring in [1b67472](https://github.com/ranocha/SummationByPartsOperators.jl/pull/268/commits/1b67472f663e8382bbf48a5d15f37dc3706c9c58). But actually I was wondering about this: Locally, it works fine for me without `using ForwardDiff`. `using SummationByPartsOperators, Optim` suffices for me to load the extension, which also confused me. Does that make sense or do you have an idea why I don't need to load ForwardDiff explicitly?

**Comment by JoshuaLampert** [comment]
Thanks a lot for your comments!
> Fair enough - feel free to add tests like this to the other files 😅

Ok, I can add some tests, but will do that in a future PR.

> > I have also noticed that the optimization needs a rather small tolerance for the gradient to be able to produce accurate results. I was also wondering about whether we should provide the user some possibility to influence the optimization, e.g., by allowing an additional kwarg options that is passed as Optim.Options to the optimization (with some reasonable default value), or by allowing to use a custom sigmoid function.

> That sounds reasonable.

I added `options` as an additional kwarg ([1b8d85e](https://github.com/ranocha/SummationByPartsOperators.jl/pull/268/commits/1b8d85e101dbdc6fce402a2947c7496b280b0c92)).

File: src/function_space_operators.jl (hunk 3 above)
Comment (ranocha) [reply src/function_space_operators.jl]: Do you have it in your Project (or the default environment)?

File: src/function_space_operators.jl (hunk 4 above)
Comment (ranocha) [reply src/function_space_operators.jl]: 👍

File: src/function_space_operators.jl (hunk 3 above)
Comment (JoshuaLampert) [reply src/function_space_operators.jl]: Yes, it was in my Project.toml (but it is not in my default environment), but I just tested that it also works without ForwardDiff being in my Project.toml. Also with ForwardDiff.jl in my Project `using SummationByPartsOperators, ForwardDiff` is not enough to call `function_space_operator`. I need `using Optim`, but not `using ForwardDiff`.

Comment (JoshuaLampert) [reply src/function_space_operators.jl]: Since Optim.jl depends on ForwardDiff.jl this makes sense. I updated the docstring accordingly.

**Comment by JoshuaLampert** [comment]
Since the optimization is doing weird things and does not work in the Julia 1.6 test, we decided to restrict this feature to Julia v1.9, see [aa868bd](https://github.com/ranocha/SummationByPartsOperators.jl/pull/268/commits/aa868bd5f65432d478b3f8925552f3a92a35adf7).

**Comment by phioeffn** [comment]
Hi, nice think! I was looking at it recently. If you have problems and comments, please let me know. We can also discuss some details. Best, Philipp

**Comment by JoshuaLampert** [comment]
Hi Philipp! Sure, I would be very interested in a discussion.

**Comment by JoshuaLampert** [comment]
In order to avoid using AD for the gradient computation, I calculated the analytical gradient of the objective function, see
[gradient_computation.pdf](https://github.com/user-attachments/files/15513567/gradient_computation.pdf). Unfortunately, the current implementation is much slower than using AD.

**Comment by JoshuaLampert** [comment]
The most part is spent in calculating `daij_dsigmak`, which can be seen by the following output from using ProfileView.jl'
s `@profview`:
Profileview
Here, I have put some line numbers to the most time-consuming parts. That is the `.jlprof` if you want to have a look on your own (I'm uploading it as a .txt because GitHub doesn't allow me to upload .jlprof files):
[profileview.txt](https://github.com/user-attachments/files/15517078/profileview.txt)

There is probably a lot of potential for performance improvements @ranocha.

File: ext/SummationByPartsOperatorsOptimForwardDiffExt.jl [hunk 6]
```diff
@@ -0,0 +1,239 @@
...
+    #     return norm(A)^2
+    # end
+    @views function optimization_function_and_grad!(F, G, x, p)
+        V, V_x, R, x_length, S_cache, A_cache, SV_cache, PV_x_cache, daij_dsigmak, daij_drhok = p
+        S = get_tmp(S_cache, x)
+        A = get_tmp(A_cache, x)
+        SV = get_tmp(SV_cache, x)
+        PV_x = get_tmp(PV_x_cache, x)
+        (N, _) = size(R)
+        L = Integer(N*(N - 1)/2)
```
Comment (ranocha) [inline ext/SummationByPartsOperatorsOptimForwardDiffExt.jl:155]:
```suggestion
        L = div(N * (N - 1), 2)
```

File: ext/SummationByPartsOperatorsOptimForwardDiffExt.jl [hunk 7]
```diff
@@ -0,0 +1,239 @@
...
+    P *= x_length / sum(P)
+    return P
+end
+
+function construct_function_space_operator(basis_functions, x_min, x_max, nodes,
+                                           ::GlaubitzNordströmÖffner2023;
+                                           options = Options(g_tol = 1e-14, iterations = 10000))
+    K = length(basis_functions)
+    N = length(nodes)
+    L = Integer(N*(N - 1)/2)
```
Comment (JoshuaLampert) [inline ext/SummationByPartsOperatorsOptimForwardDiffExt.jl:103]:
```suggestion
    L = div(N * (N - 1), 2)
```

File: ext/SummationByPartsOperatorsOptimForwardDiffExt.jl [hunk 8]
```diff
@@ -0,0 +1,239 @@
...
+        if !isnothing(G)
+            fill!(daij_dsigmak, zero(eltype(daij_dsigmak)))
+            for i in 1:N
+                for j in 1:K
+                    for k in 1:L
+                        l_tilde = k + i - N * (i - 1) + div(i * (i - 1), 2)
+                        # same as above, but needs more type conversions
+                        # l_tilde = Int(k + i - (i - 1) * (N - i/2))
+                        if i + 1 <= l_tilde <= N
+                            daij_dsigmak[i, j, k] += V[l_tilde, j]
```
Comment (ranocha) [inline ext/SummationByPartsOperatorsOptimForwardDiffExt.jl:172]: Can you change the order of the loops? Julia is column-major, so the `i` index of `daij_dsigmak` should be updated closely together

File: ext/SummationByPartsOperatorsOptimForwardDiffExt.jl [hunk 9]
```diff
@@ -0,0 +1,239 @@
...
+                            daij_drhok[i, j, k] = factor * sig_rho[i]
+                        end
+                    end
+                end
+            end
+            for k in 1:L
+                G[k] = 2 * sum(daij_dsigmak[:, :, k] .* A)
+            end
+            for k in 1:N
+                G[L + k] = 2 * sum(daij_drhok[:, :, k] .* A)
```
Comment (ranocha) [inline ext/SummationByPartsOperatorsOptimForwardDiffExt.jl:212]: These lines allocate, don't they? Aren't they the same as
```suggestion
                G[k] = 2 * dot(daij_dsigmak[:, :, k], A)
            end
            for k in 1:N
                G[L + k] = 2 * dot(daij_drhok[:, :, k], A)
```

**Comment by ranocha** [comment]
There is a lot of red stuff - did you check type stability?

File: ext/SummationByPartsOperatorsOptimForwardDiffExt.jl (hunk 8 above)
Comment (JoshuaLampert) [reply ext/SummationByPartsOperatorsOptimForwardDiffExt.jl]: Done in [c133295](https://github.com/ranocha/SummationByPartsOperators.jl/pull/268/commits/c133295f4e1cab0c79efe2ba92c34cb232d6e7dd).

**Comment by JoshuaLampert** [comment]
> There is a lot of red stuff - did you check type stability?

No, you mean with Cthulhu.jl for example, right? I should probably do that.

**Comment by ranocha** [comment]
Did it change anything?

**Comment by JoshuaLampert** [comment]
It's something like 15 % faster. I think especially because of the `dot` instead of `sum` change. But interestingly, we also have less iterations after that change. But I guess this is just because of some small floating point differences and because we use so small tolerances. The profile didn't change much, however. There is still a lot of red on the top.

**Comment by ranocha** [comment]
Can you extract `optimization_function_and_grad!` from `construct_function_space_operator` and check its type stability? To me, it looks like there is some type stability happening there.

**Comment by ranocha** [comment]
You should move `optimization_function_and_grad!` outside of `construct_function_space_operator` and add `K` to the parameters `p` (and extract it in `optimization_function_and_grad!`). I think this should improve the performance

**Comment by ranocha** [comment]
I just did that and also used `axes` instead of `1:K` etc. Could you please check whether the results still look reasonable to you and improve the performance?

**Comment by JoshuaLampert** [comment]
Oh, now you were a few seconds faster than me, but yes that was it! No it is only a few milliseconds compared to 3 seconds before :sweat_smile:. Thanks a lot @ranocha!

File: ext/SummationByPartsOperatorsOptimForwardDiffExt.jl [hunk 10]
```diff
@@ -0,0 +1,239 @@
...
+                    # l_tilde = Int(k + i - (i - 1) * (N - i/2))
+                    if i + 1 <= l_tilde <= N
+                        daij_dsigmak[i, j, k] += V[l_tilde, j]
+                    else
+                        C = N^2 - 3*N + 2*i - 2*k + 1/4
+                        if C >= 0
+                            D = sqrt(C)
+                            if isinteger(1/2 + D)
+                                l_hat1 = Int(N - 1/2 + D)
+                                l_hat2 = Int(N - 1/2 - D)
```
Comment (ranocha) [inline ext/SummationByPartsOperatorsOptimForwardDiffExt.jl:200]: If you want to optimize it further, this is probably the part to look at. Can you get the same results using only integer operations?

If not, maybe you can also reuse some results, e.g., 

```suggestion
                        C = N^2 - 3*N + 2*i - 2*k + 1/4
                        if C >= 0
                            D = sqrt(C)
                            D_plus_one_half = D + 0.5
                            D_plus_one_half_trunc = trunc(D_plus_one_half)
                            if D_plus_one_half == D_plus_one_half_trunc
                                int_D_plus_one_half = trunc(Int, D_plus_one_half_trunc)
                                l_hat1 = N + int_D_plus_one_half - 1
                                l_hat2 = N - int_D_plus_one_half
```

**Comment by JoshuaLampert** [comment]
Now it is also much faster than with ForwardDiff.jl, especially for high number of points. Nice!

**Comment by ranocha** [comment]
Cool 😎

File: ext/SummationByPartsOperatorsOptimForwardDiffExt.jl (hunk 10 above)
Comment (JoshuaLampert) [reply ext/SummationByPartsOperatorsOptimForwardDiffExt.jl:200]: Using only integer arithmetic will probably be hard because of the `sqrt`, right? I don't see how we could get rid of that, but let me try your suggestion with reusing some computations.

Comment (JoshuaLampert) [reply ext/SummationByPartsOperatorsOptimForwardDiffExt.jl]: This didn't change so much.

**Comment by JoshuaLampert** [comment]
It is much faster now, but `N = 50` nodes it still takes ~8.8 seconds on my laptop to create the operators. This scales pretty poorly in `N` (which is expected since `daij_drhok` is of size $N \cdot L \cdot K = O(N^3\cdot K)$ ).

File: ext/SummationByPartsOperatorsOptimForwardDiffExt.jl [hunk 11]
```diff
@@ -0,0 +1,222 @@
...
+using SparseArrays: spzeros
+
+function SummationByPartsOperators.function_space_operator(basis_functions,
+                                                           x_min::T, x_max::T, nodes::Vector{T},
+                                                           source::SourceOfCoefficients;
+                                                           derivative_order = 1, accuracy_order = 0,
+                                                           options = Options(g_tol = 1e-14, iterations = 10000)) where {T, SourceOfCoefficients}
+
+    @assert derivative_order == 1 "Only first derivative operators are supported"
+    sort!(nodes)
```
Comment (JoshuaLampert) [inline ext/SummationByPartsOperatorsOptimForwardDiffExt.jl:18]: I'm not sure if sorting the nodes is reasonable or necessary. It also works without sorting, but having a non-sorted grid in the `MatrixDerivativeOperator` feels a bit strange to me. What do you think, @ranocha?

File: ext/SummationByPartsOperatorsOptimForwardDiffExt.jl [hunk 12]
```diff
@@ -0,0 +1,222 @@
...
+using Optim: Optim, Options, LBFGS, optimize, minimizer
+using ForwardDiff: ForwardDiff
+
+using SummationByPartsOperators: SummationByPartsOperators, GlaubitzNordströmÖffner2023, MatrixDerivativeOperator
+using LinearAlgebra: Diagonal, LowerTriangular, dot, diag, norm, mul!
+using PreallocationTools: DiffCache, get_tmp
+using SparseArrays: spzeros
+
+function SummationByPartsOperators.function_space_operator(basis_functions,
+                                                           x_min::T, x_max::T, nodes::Vector{T},
```
Comment (JoshuaLampert) [inline ext/SummationByPartsOperatorsOptimForwardDiffExt.jl:12]: I'm also not sure how to deal with `x_min` and `x_max`. We could

1. Keep it as it is, which means that if `x_min` and/or `x_max` are not within `nodes`, then the grid of the `MatrixDerivativeOperator` is not the same as the `nodes`, which feels unintuitive, when you think of `D` as global operator and not an operator on a reference element.
2. Error if `x_min` and/or `x_max` are not within `nodes`.
3. If they are not within `nodes` implicitly add them, which also doesn't feel quite correct.
4. Remove the arguments `x_min` and `x_max` and just pass `first(nodes)` and `last(nodes)` to the `MatrixDerivativeOperator`.

I think I tend towards option 4. What is your opinion, @ranocha?

**Comment by JoshuaLampert** [comment]
Do you have any other reasonable tests in mind, I could add, @ranocha?

**Comment by JoshuaLampert** [comment]
I would also like to stress that the optimization can be very vulnerable. Consider the following example:

```julia-repl
julia> using SummationByPartsOperators, Optim

julia> nodes = [0.0, rand(3)..., 1.0]
5-element Vector{Float64}:
 0.0
 0.05651765804974351
 0.29863912793993175
 0.657232875727107
 1.0

julia> basis_functions = [one, identity, exp]
3-element Vector{Function}:
 one (generic function with 45 methods)
 identity (generic function with 1 method)
 exp (generic function with 21 methods)

julia> D = function_space_operator(basis_functions, first(nodes), last(nodes), nodes, GlaubitzNordströmÖffner2023())
Matrix-based first-derivative operator {T=Float64} on 5 nodes in [0.0, 1.0]

julia> D * ones(5)
5-element Vector{Float64}:
 -2038.0570044219494
    -1.073365630488432e-6
     6.732198513992316e-6
     3.664382162149593e-6
    -2.1692864620170837e-5
```

which is _very_ far from `zeros(5)`. The optimization reports:

```julia
 * Status: failure (reached maximum number of iterations)

 * Candidate solution
    Final objective value:     2.677287e-10

 * Found with
    Algorithm:     L-BFGS

 * Convergence measures
    |x - x'|               = 8.05e-04 ≰ 0.0e+00
    |x - x'|/|x'|          = 3.85e-05 ≰ 0.0e+00
    |f(x) - f(x')|         = 6.90e-22 ≰ 0.0e+00
    |f(x) - f(x')|/|f(x')| = 2.58e-12 ≰ 0.0e+00
    |g(x)|                 = 1.17e-11 ≰ 1.0e-14

 * Work counters
    Seconds run:   0  (vs limit Inf)
    Iterations:    10000
    f(x) calls:    26276
    ∇f(x) calls:   26276
```

So even with an objective value of ~2.7e-10, which is quite small, the operator is very far from being exact. Increasing the number of iterations significantly and decreasing the tolerance doesn't help:

```julia-repl
julia> D = function_space_operator(basis_functions, first(nodes), last(nodes), nodes, GlaubitzNordströmÖffner2023(); options = Optim.Options(g_tol = 1e-16, iterations = 200000))
 * Status: success

 * Candidate solution
    Final objective value:     2.677287e-10

 * Found with
    Algorithm:     L-BFGS

 * Convergence measures
    |x - x'|               = 2.01e-07 ≰ 0.0e+00
    |x - x'|/|x'|          = 5.40e-09 ≰ 0.0e+00
    |f(x) - f(x')|         = 9.38e-23 ≰ 0.0e+00
    |f(x) - f(x')|/|f(x')| = 3.50e-13 ≰ 0.0e+00
    |g(x)|                 = 8.08e-17 ≤ 1.0e-16

 * Work counters
    Seconds run:   0  (vs limit Inf)
    Iterations:    102609
    f(x) calls:    270140
    ∇f(x) calls:   270140

Matrix-based first-derivative operator {T=Float64} on 5 nodes in [0.0, 1.0]

julia> D * ones(5)
5-element Vector{Float64}:
 -2.46283614165e10
 -1.0733685033015306e-6
  6.7321802603159675e-6
  3.664360717303694e-6
 -2.1692791865035588e-5
```

I think this is a general issue with the optimization process and thus I think a deeper theoretical insight in the optimization would be useful.
To sum up, if we want to merge this, I would suggest to put a warning in the docstring that this is still an experimental feature and might change in the future.

File: ext/SummationByPartsOperatorsOptimForwardDiffExt.jl (hunk 11 above)
Comment (ranocha) [reply ext/SummationByPartsOperatorsOptimForwardDiffExt.jl:18]: I prefer sorting the nodes for consistency with the other operators 👍

File: ext/SummationByPartsOperatorsOptimForwardDiffExt.jl (hunk 12 above)
Comment (ranocha) [reply ext/SummationByPartsOperatorsOptimForwardDiffExt.jl:12]: Yes, that sounds reasonable. We should mention this in the docstring.

**Comment by ranocha** [comment]
> To sum up, if we want to merge this, I would suggest to put a warning in the docstring that this is still an experimental feature and might change in the future.

Yes, that's a very good suggestion 👍

**Comment by ranocha** [comment]
> I would also like to stress that the optimization can be very vulnerable.

It looks like this is actually a question for further research and not only an implementation question.

**Comment by JoshuaLampert** [comment]
> It looks like this is actually a question for further research and not only an implementation question.

Yes, definitely. I was just trying to compute the Hessian of the function to check convexity.

File: ext/SummationByPartsOperatorsOptimForwardDiffExt.jl [hunk 13]
```diff
@@ -0,0 +1,222 @@
+module SummationByPartsOperatorsOptimForwardDiffExt
+
+using Optim: Optim, Options, LBFGS, optimize, minimizer
+using ForwardDiff: ForwardDiff
+
+using SummationByPartsOperators: SummationByPartsOperators, GlaubitzNordströmÖffner2023, MatrixDerivativeOperator
+using LinearAlgebra: Diagonal, LowerTriangular, dot, diag, norm, mul!
+using PreallocationTools: DiffCache, get_tmp
```
Comment (ranocha) [inline ext/SummationByPartsOperatorsOptimForwardDiffExt.jl:8]: Are these still required after implementing the gradient analytically?

File: test/matrix_operators_test.jl [hunk 14]
```diff
@@ -1,6 +1,7 @@
 using Test
 using LinearAlgebra
 using SummationByPartsOperators
+using Optim
```
Comment (ranocha) [inline test/matrix_operators_test.jl:4]:
```suggestion
using Optim: Optim # to enable loading the function space operator optimization code
```

**Comment by ranocha** [comment]
> Do you have any other reasonable tests in mind, I could add, @ranocha?

The tests look reasonable 👍

File: ext/SummationByPartsOperatorsOptimForwardDiffExt.jl [hunk 15]
```diff
@@ -0,0 +1,222 @@
...
+                        if C >= 0
+                            D = sqrt(C)
+                            D_plus_one_half = D + 0.5
+                            D_plus_one_half_trunc = trunc(D_plus_one_half)
+                            if D_plus_one_half == D_plus_one_half_trunc
+                                int_D_plus_one_half = trunc(Int, D_plus_one_half_trunc)
+                                l_hat1 = N + int_D_plus_one_half - 1
+                                l_hat2 = N - int_D_plus_one_half
+                                if 1 <= l_hat1 <= i - 1
+                                    daij_dsigmak[i, j, k] -= V[l_hat1, j]
```
Comment (ranocha) [inline ext/SummationByPartsOperatorsOptimForwardDiffExt.jl:183]: This line is reported as not being covered by tests. Any idea why? Is it needed or is this case impossible?

**Comment by JoshuaLampert** [comment]
> > To sum up, if we want to merge this, I would suggest to put a warning in the docstring that this is still an experimental feature and might change in the future.
> 
> Yes, that's a very good suggestion 👍

Done in [38081b1](https://github.com/ranocha/SummationByPartsOperators.jl/pull/268/commits/38081b1822de3efbadef4a23afe5e6dcacd87155).

File: ext/SummationByPartsOperatorsOptimForwardDiffExt.jl (hunk 15 above)
Comment (JoshuaLampert) [reply ext/SummationByPartsOperatorsOptimForwardDiffExt.jl]: Yes, this case is indeed impossible: We have $\hat{l}_1 = N - 1/2 + \sqrt{C} > N - 1 \ge i - 1$. Thanks for the hint! See [fd470ee](https://github.com/ranocha/SummationByPartsOperators.jl/pull/268/commits/fd470ee623f3a9585a5be3c6558fd2e99cf9af98).

File: ext/SummationByPartsOperatorsOptimForwardDiffExt.jl (hunk 12 above)
Comment (ranocha) [reply ext/SummationByPartsOperatorsOptimForwardDiffExt.jl]: Thanks!

File: ext/SummationByPartsOperatorsOptimForwardDiffExt.jl (hunk 15 above)
Comment (ranocha) [reply ext/SummationByPartsOperatorsOptimForwardDiffExt.jl]: Thanks

File: ext/SummationByPartsOperatorsOptimForwardDiffExt.jl (hunk 13 above)
Comment (JoshuaLampert) [reply ext/SummationByPartsOperatorsOptimForwardDiffExt.jl:8]: Ah no, I don't think so since I introduced them to be able to apply automatic differentiation. I'll remove PreallocationTools again.

Comment (JoshuaLampert) [reply ext/SummationByPartsOperatorsOptimForwardDiffExt.jl]: I removed it again in [7f54a30](https://github.com/ranocha/SummationByPartsOperators.jl/pull/268/commits/7f54a303cfcb0c0f5be49700223c546de6ce83e3).

**Review by ranocha — APPROVED** [review]
Thanks!

**Comment by JoshuaLampert** [comment]
FYI @phioeffn @jglaubitz
